Seating charts - 2024 edition

carl-vbn opened the pull request:
This PR is here to make things a lot easier to work with

- Added `imagedl.py` script to facilitate image downloading
- Added `roster.py` to download rosters. It is also imported by other scripts to standardize roster parsing from CSV files (it can reliably load csv files with or without canvas headers)
- `go_brr` now imports `seatingchart` instead of calling it with `os.system`
- Updated README.md
- A counter of selected seats is now shown on the HTML pages, to make headcounts easier

fixes #32 #26 and starts on #29

carl-vbn committed:
Image downloader script and more robust roster loading

carl-vbn committed:
Cleaned up go_brr.py

carl-vbn committed:
Reconciled go_brr with previous commits I forgot to pull

carl-vbn committed:
Fixed seatingchart.py

carl-vbn committed:
Linked go_brr and seatingchart

carl-vbn committed:
Added ability to download roster from canvas

carl-vbn committed:
Updated README.md

carl-vbn committed:
Updated README.md (again)

carl-vbn committed:
Added ASCII art to go_brr.py

carl-vbn committed:
Added selected seat counter indicator to HTML

carl-vbn committed:
Added layout format documentation to README

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+        url = url_prefix+uni+'.jpg'
+        
+        counter += 1
+        percentage = f"{round(counter/uni_count*100, 1)}%"
```
Could also just calculate the percentage as a number and put in directly in the print below instead of formatting as a string

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+        
+        counter += 1
+        percentage = f"{round(counter/uni_count*100, 1)}%"
+        print("Downloading", url, f"({counter}/{uni_count} -- {percentage})")
```
Would recommend either overwriting the previous line to avoid spamming terminal output or just using tqdm, but not crucial

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+        print("Downloading", url, f"({counter}/{uni_count} -- {percentage})")
+        response = requests.get(url, headers=headers)
+
+        with open(os.path.join(output_dir, uni+'.jpg'), 'wb') as f:
+            f.write(response.content)
```
I'd use pathlib over os.path for path stuff -- could also use pathlib.write_bytes() directly here

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+
+    print(f'[{last_step_number}]', instruction)
+    last_step_number += 1
+    input()
```
`input()` takes a prompt, can just put the print contents in that using an f-string

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+    print(TITLE_ART)
+    print("Python utility to download student images automatically")
+
+    if not os.path.exists(output_dir):
+        print("Output directory does not exist. Creating...")
+        os.mkdir(output_dir)
```
Would recommend using pathlib

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+        unis = set(map(lambda student: student[0], students))
+
+        if skip_existing:
+            skipped_unis = set(map(lambda file: file.split('.')[0], os.listdir(output_dir))).intersection(unis)
```
I would use a set comprehension over `set(map())` here if possible: `{file.split('.')[0] for file in os.listdir(output_dir)}` should work (can also use pathlib's iterdir instead)

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+
+    print(YELLOW + "The instructions were designed for chromium based browsers and might differ on other browsers.\n" + END)
+    print(LIGHT_BLUE + "This utility will guide you through a series of steps. Once you complete a step, press ENTER" + END)
+    step("Press ENTER to begin")
+    step("Open canvas/courseworks and go to the relevant course, then open your browser's developer tools. Find the network tab and press 'clear'")
+    step("Navigate to 'Photo Roster' section on the course canvas page and wait for images to load (this can take a while)")
+    step("The network tab should fill with requests to images named named after student's unis. Right click one of these requests and select 'Copy as cURL (bash)'. Be careful NOT to select the option saying 'Copy ALL'")
```
I'm not sure how necessary it is to make the user press enter after each step, but for now it's fine -- easy enough to change if necessary

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+    curl = multiline_input()
+
+    url, headers = parse_curl(curl)
+    url_prefix = "/".join(url.split('/')[:-1])+'/'
```
Add space between + sign
A comment giving an example of the transformation this is doing would be good (i.e. "Ex: A -> B""

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+    if input("Are you ready to begin? (Y/n) ") not in ['Y','y','']:
+        print('Aborted.')
+        exit(0)
+
```
Remove one blank line

tzussman commented on imagedl.py:
```diff
@@ -0,0 +1,136 @@
+    last_step_number += 1
+    input()
+
+def run_guide(output_dir, unis=None, roster_filepath=None, skip_existing=False):
```
Does run_guide ever get called with unis set? Why do we need it?

tzussman commented on rosters.py:
```diff
@@ -0,0 +1,103 @@
+    response = requests.get(f'https://courseworks2.columbia.edu/api/v1/courses/{course_id}/students',
+            headers={'Authorization': f'Bearer {canvas_api_key}'}).json()
+
+    response = list(filter(lambda x: x['sortable_name'] != 'Student, Test', response))
```
I'm not a big fan of the `list(filter())` style, would prefer a conditional list comprehension instead, but up to you. Same for the instances below

tzussman commented on seatingchart.py:
```diff
@@ -8,6 +8,7 @@
 import csv
 import itertools
 import random
+import rosters
```
Same comment about import style as before

tzussman commented on seatingchart.py:
```diff
@@ -162,16 +162,15 @@ def main(args):
-        for seat, uni in assignments.items():
-            csv_output.writerow(list(uni)[:2] + [seat])
+        for seat, student in assignments.items():
+            csv_output.writerow(list(student) + [seat])
```
Can probably just do [student, seat] (unless I'm missing something)

tzussman commented on rosters.py:
```diff
@@ -0,0 +1,103 @@
+    else:
+        return elements[:-1]
+
+
```
Would be good to have documentation about what format this expects the roster to be in and what format it returns it in -- those have probably been the main source of issues with this repo over the past couple years

tzussman commented on rosters.py:
```diff
@@ -0,0 +1,103 @@
+            with open(args.download, 'w') as f:
+                csv.writer(f).writerows(students)
+
+            print(GREEN+"Roster saved successfully."+END)
+        except KeyboardInterrupt:
+            print(RED+"Download cancelled."+END)
+        except Exception:
+            print(RED+"Failed to download roster. Please check your API key and course ID"+END)
```
Add spaces

tzussman reviewed (changes requested):
Overall this looks really good! I've left a bunch of small comments throughout.
My main concern at the moment is rosters.py -- I didn't take too close of a look at it, but I'm worried that it duplicates functionality that we have in get_class.py in cloud-mgt. More importantly, it looks like it uses a different API endpoint and does some more parsing than get_class does, so I just want to make sure that there aren't some small issues we're missing there (actually it looks like it does it better than get_class so there may be some improvements to make there). It could also do with some more documentation about what format it expects as input and the structure of the roster it returns in Python.

carl-vbn committed:
Removed extra hashes after section names in README

carl-vbn committed:
Added newlines in README.md

carl-vbn committed:
Added spaces when concatenating ANSI color escape sequences

carl-vbn committed:
Import ordering readjusted

carl-vbn committed:
Removed redundant spaces in imagedl.py

carl-vbn committed:
imagedl cleanup

chmullig commented on seatingchart.py:
```diff
@@ -255,7 +263,7 @@ def run(slug, layout, title=None, lefty=False, debug=False):
 
             seating_chart.write("</tr>\n\n")
-        seating_chart.write("</table></body>")
+        seating_chart.write('</table><p class="selected-count-line"><span id="selected-counter">0</span> seats selected.</p></body>')
```
This is so clever. Wish I'd thought of that!

carl-vbn committed:
rosters.py cleanup and better explanation

carl-vbn committed:
seatingchart.py cleanup

carl-vbn committed:
Fixed bug introduced in rosters.py cleanup

carl-vbn committed:
Fixed error introduced in seatingchart.py

carl-vbn committed:
imagedl.py improvements

carl-vbn committed:
pathlib migration

carl-vbn committed:
Simplified file writing using write_bytes

carl-vbn commented:
@tzussman I implemented pretty much all of your feedback.

For rosters.py, I see your point. I added that file because it is very straightforward, not that many moving parts and allows the entire repo to be self-sufficient from retrieving all the files to generating the html. The advantage of the endpoint I'm using is that it returns the entire class in one query (I double checked, the number of entries it gives matches the number of valid rows in the CSV exported from the Grades tab).

If you think we should rely on cloud-mgt instead, I can easily strip the downloading functionality out of it. The file is also used by the other scripts to parse roster files, but none of them use the downloading functionality, that's fully manual. I've also added some more details on why this file is useful vs just loading the CSV in the other files directly.

tzussman reviewed (commented):
Left a couple more small comments – this is almost there!
Re: rosters.py, I think it's fine to use it here. In fact, I think you should look into changing the cloud-mgt version to use the endpoint you found, since it seems much simpler.

carl-vbn committed:
Removed unused os import

carl-vbn committed:
Added comment about Test Student

carl-vbn committed:
roster rework

carl-vbn committed:
Print roster size after downloading

carl-vbn committed:
Removed unused variable

carl-vbn committed:
Added comment about roster format to rosters.py

carl-vbn committed:
Removed debugging message

carl-vbn commented:
@tzussman I hope this covers it! I also made rosters.py print things out a bit more nicely using a table, because it was just printing a big blob of unis before.

tzussman reviewed (approved):
This looks great, thanks Carl! Assuming you tested the latest round of updates, feel free to go ahead and merge! (I didn't closely read the `print_table()` code, but I trust that it looks a lot better now)

carl-vbn commented:
I tested it all with the last changes before committing, so it should be good to go then. Thanks for all your feedback Tal!
